packages/wallet/src/redux/channelState/reducer.ts

```diff
@@ -20,22 +7,105 @@ import { challengingReducer } from './challenging/reducer';
 import { respondingReducer } from './responding/reducer';
 import { withdrawingReducer } from './withdrawing/reducer';
 import { closingReducer } from './closing/reducer';
-import { WalletAction, CONCLUDE_REQUESTED, COMMITMENT_RECEIVED } from '../actions';
+import {
+  WalletAction,
+  CONCLUDE_REQUESTED,
+  COMMITMENT_RECEIVED,
+  CHANNEL_INITIALIZED,
+  ChannelAction,
+  isReceiveFirstCommitment,
+} from '../actions';
 import {
   unreachable,
   ourTurn,
   validTransition,
   ReducerWithSideEffects,
+  combineReducersWithSideEffects,
 } from '../../utils/reducer-utils';
 import { validCommitmentSignature } from '../../utils/signing-utils';
-import { showWallet } from 'magmo-wallet-client/lib/wallet-events';
+import { showWallet, channelInitializationSuccess } from 'magmo-wallet-client/lib/wallet-events';
 import { CommitmentType } from 'fmg-core';
 import { StateWithSideEffects } from '../shared/state';
+import { ethers } from 'ethers';
+import { channelID } from 'fmg-core/lib/channel';
 
-export const channelReducer: ReducerWithSideEffects<ChannelState> = (
-  state: ChannelState,
+export const channelStateReducer: ReducerWithSideEffects<states.ChannelState> = (
+  state: states.ChannelState,
   action: WalletAction,
-): StateWithSideEffects<ChannelState> => {
+): StateWithSideEffects<states.ChannelState> => {
+  if (isReceiveFirstCommitment(action)) {
+    // We manually select and move the initializing channel into the initializedChannelState
+    // before applying the combined reducer, so that the address and private key is in the
+    // right slot (by its channelId)
+    const channel = action.commitment.channel;
+    const channelId = channelID(channel);
+    if (state.initializedChannels[channelId]) {
+      throw new Error('Channel already exists');
+    }
+    const initializingAddresses = new Set(Object.keys(state.initializingChannels));
+    const ourAddress = channel.participants.find(addr => initializingAddresses.has(addr));
+    if (!ourAddress) {
+      return { state };
+    }
+    const ourIndex = channel.participants.indexOf(ourAddress);
+
+    const { address, privateKey } = state.initializingChannels[ourAddress];
+    delete state.initializingChannels[ourAddress];
```

Comment:
The state object is being modified here. Not sure if that is dangerous given the Redux paradigms.